### DISCUSSION TITLE
Built-in NGINX reverse proxy for local connections

### DISCUSSION THREAD

**Discussion by felipecrs** [homeassistant-apps/app-cloudflared#292] [Ideas]
I use my own DNS server in my local connection that has a different name resolution for my HA external address. For example:

my-ha.com through 8.8.8.8 -> resolves to Cloudflare IP
my-ha.com through 192.168.1.10 -> resolves to my internal Home Assistant IP

In order to make that work, when connecting locally, the internal Home Assistant has to expose itself on port 443 and also serve a valid certificate.

I use:

1. Let's Encrypt add-on to generate and renew this local certificate automatically
2. NGINX SSL Proxy add-on to expose HA through 443, so that I can continue to connect to my HA locally with  http://homeassistant.local:8123

But there are some caveats with this:

1. What if I have additional hostnames? They would not resolve internally because only NGINX SSL Proxy will be running on port 443 and I cannot make something else run on port 443. Let's say Traccar for example, which uses port 8082 by default:
   - my-traccar.com through 8.8.8.8 resolves to Cloudflare's IP (OK)
   - my-traccar.com through 192.168.1.10 resolves to my HA IP (OK, but won't work because port 443 is not Traccar's port)
   
I know that I can fix this issue by using NGINX Proxy Manager and setup everything by myself. But I also think it would be astonishing if Cloudflared add-on could come with its own NGINX server that can optionally run on port 443 and would forward everything according to the configuration set in _Additional Hosts_, including HA itself so I don't need NGINX SSL Proxy. Imagine if I could drop the dependency on Let's Encrypt add-on even? Cloudflared could support generating certbot certificates by using the Cloudflare DNS Challenge for an integrated experience.

Anyway, what do you guys think?

**Comment by brenner-tobias** [comment]
Hi @felipecrs,
thanks for this suggestion. The idea sounds good to have such a full-blown solution, but I do think that having a more distributed and specialised approach is the better way to move forward here. The add-on would become very big and complicated and we do have specialised add-ons for your use-case, so I do not see any harm in using and combining them.

Having that said, I am open for any other suggestions / thoughts. @elcajon what is your perspective here?

**Reply by felipecrs** [reply]
After putting more thought on it, I think I agree with you. I'm now using NGINX Proxy Manager, and it works cool. Of course I have more setup to do, but it's a one-time thing.

By the way, I'm not using the _Enable Catch-All Nginx-Proxy-Manager_ option for Cloudflared. And I don't know whether I should or not. Things seems to be working well without it though.

**Reply by brenner-tobias** [reply]
ok, thats great to hear.
Regarrding the catch-all option: If you manually add all the hosts that you want to route from Cloudflared into the tunnel to NPM in the additional_hosts option, this works as well. The flag just signals cloudflared tunnel to route everything, that does not meet any other rule, to NPM instead of dropping it. So both approaches work.

**Reply by felipecrs** [reply]
Got it. Thanks!

**Reply by felipecrs** [reply]
Now that we already have NGINX running inside the add-on, can we revisit this?

What I think would need to be done is:

1. `additional_hosts` would go through NGINX too
1. Exposure of NGINX's port 80 and 443 to outside the container
1. Let's Encrypt for generating local certificates for NGINX

### Discussions

`cloudflared` could also communicate with NGINX through https. This would allow [Strict SSL](https://developers.cloudflare.com/ssl/origin-configuration/ssl-modes/full-strict/) mode.

Perhaps replacing NGINX with [pingap](https://github.com/vicanso/pingap), which is based on Cloudflare's Pingora, and should be more performant and more lightweight.  Unlike NGINX, pingap provides built-in Let's Encrypt functionality, which may come in handy.

**Reply by felipecrs** [reply]
Maybe pingap is too experimental for the task. [Caddy](https://github.com/caddyserver/caddy) would be another great alternative, with built-in support for Let's Encrypt.

The nice thing of being connected with `cloudflared` is that it should allow the HTTP challenge to work for generating certificates, without even requiring more credentials from users.

**Reply by felipecrs** [reply]
@brenner-tobias, @elcajon, I can work on it, unless any objection.

**Reply by elcajon** [reply]
I understand the intention and why it would be useful. I also like Caddy because it is pretty easy to setup and issue certificates, but I am not sure how to deal with it, without a breaking change for the current user base.

Just some first thoughts:
- Do we just expose 80 and 443 or run the add-on on host network?
- What do we do with the current nginx catch-all function? Just keep it?
- How to migrate current logic without breaking anything

Maybe we should start with some kind of PoC?

**Reply by felipecrs** [reply]
Right. I will start the PoC, and we can continue from there. I am confident I can provide a clean solution that addresses these concerns.

**Reply by brenner-tobias** [reply]
Fine for me, lets keep it simple and focus on a clean code-base.

**Reply by ericyew** [reply]
Curious as to where this got to? Was this ever implemented into a PoC or a functional solution?

**Reply by brenner-tobias** [reply]
The solution was implemented by @elcajon and released in [v5.2.8](https://github.com/brenner-tobias/addon-cloudflared/releases/tag/v5.2.8) and improved since then.
Right now, we are working on changing to Caddy, which is driven by @felipecrs in #843
